boot/bootutil/src/loader.c

```diff
@@ -1247,116 +1297,104 @@ boot_go(struct boot_rsp *rsp)
-     */
-    if (boot_slots_compatible()) {
-        rc = boot_swap_if_needed(&swap_type);
-        assert(rc == 0);
+        /* Attempt to read an image header from each slot. */
+        rc = boot_read_image_headers();
         if (rc != 0) {
             goto out;
         }
 
-        /*
-         * The following states need image_ok be explicitly set after the
-         * swap was finished to avoid a new revert.
+        /* If the image slots aren't compatible, no swap is possible.  Just boot
+         * into slot 0.
          */
-        if (swap_type == BOOT_SWAP_TYPE_REVERT || swap_type == BOOT_SWAP_TYPE_FAIL) {
-            rc = boot_set_image_ok();
+        if (boot_slots_compatible()) {
+            rc = boot_swap_if_needed(&swap_type);
+            assert(rc == 0);
             if (rc != 0) {
-                swap_type = BOOT_SWAP_TYPE_PANIC;
+                goto out;
             }
-        }
-    } else {
-        swap_type = BOOT_SWAP_TYPE_NONE;
-    }
 
-    switch (swap_type) {
-    case BOOT_SWAP_TYPE_NONE:
-        slot = 0;
-        break;
-
-    case BOOT_SWAP_TYPE_TEST:          /* fallthrough */
-    case BOOT_SWAP_TYPE_PERM:          /* fallthrough */
-    case BOOT_SWAP_TYPE_REVERT:
-        slot = 1;
-        reload_headers = true;
-        rc = boot_set_copy_done();
-        if (rc != 0) {
-            swap_type = BOOT_SWAP_TYPE_PANIC;
+            /*
+             * The following states need image_ok be explicitly set after the
+             * swap was finished to avoid a new revert.
+             */
+            if (swap_type == BOOT_SWAP_TYPE_REVERT || swap_type == BOOT_SWAP_TYPE_FAIL) {
+                rc = boot_set_image_ok();
+                if (rc != 0) {
+                    swap_type = BOOT_SWAP_TYPE_PANIC;
+                }
+            }
+        } else {
+            swap_type = BOOT_SWAP_TYPE_NONE;
         }
-        break;
 
-    case BOOT_SWAP_TYPE_FAIL:
-        /* The image in slot 1 was invalid and is now erased.  Ensure we don't
-         * try to boot into it again on the next reboot.  Do this by pretending
-         * we just reverted back to slot 0.
-         */
-        slot = 0;
-        reload_headers = true;
-        break;
-
-    default:
-        swap_type = BOOT_SWAP_TYPE_PANIC;
-    }
-
-    if (swap_type == BOOT_SWAP_TYPE_PANIC) {
-        BOOT_LOG_ERR("panic!");
-        assert(0);
-
-        /* Loop forever... */
-        while (1) {}
-    }
-
-#ifdef MCUBOOT_VALIDATE_SLOT0
-    if (reload_headers) {
+        boot_finalize_swap(swap_type, &reload_headers);
+        if (reload_headers) {
             rc = boot_read_image_headers();
             if (rc != 0) {
-                    goto out;
+                goto out;
             }
-    }
+        }
 
-    rc = boot_validate_slot(0);
-    assert(rc == 0);
-    if (rc != 0) {
-        rc = BOOT_EBADIMAGE;
-        goto out;
-    }
-#else
-    (void)reload_headers;
+#ifdef MCUBOOT_VALIDATE_SLOT0
+        rc = boot_validate_slot(0);
```

Comment:
Same here, not clear how this can be correct without specifying which image's slot 0 is being validated.

Reply:
Ah, I see this is using a global variable now... this should be changed